File: vel_harness/middleware/subagents.py
"""
Subagents Middleware

Provides tools for spawning and managing parallel subagents.
Integrates with AgentRegistry for typed agent configurations.
"""

from typing import Any, Dict, List, Optional

from vel import ToolSpec

from vel_harness.agents import AgentRegistry
from vel_harness.middleware.base import BaseMiddleware
from vel_harness.subagents.spawner import (
    SubagentConfig,
    SubagentResult,
    SubagentSpawner,
    SubagentStatus,
)


class SubagentsMiddleware(BaseMiddleware):
    """
    Middleware providing subagent spawning tools.

    Provides tools:
    - spawn_subagent: Create a new subagent for a task
    - spawn_parallel: Spawn multiple subagents for parallel research
    - wait_subagent: Wait for a subagent to complete
    - get_subagent_result: Get result from a completed subagent
    - list_subagents: List all active and completed subagents
    - cancel_subagent: Cancel a running subagent

    Subagents are independent agents that can:
    - Execute research tasks in parallel
    - Have their own tools and context
    - Report results back to the parent agent
    """

    def __init__(
        self,
        default_model: Optional[Dict[str, str]] = None,
        default_system_prompt: Optional[str] = None,
        default_tools: Optional[List[Any]] = None,
        max_concurrent: int = 5,
        max_turns: int = 10,
        timeout: float = 300.0,
        agent_registry: Optional[AgentRegistry] = None,
    ) -> None:
        """
        Initialize subagents middleware.

        Args:
            default_model: Default model config for subagents
            default_system_prompt: Default system prompt for subagents
            default_tools: Default tools for subagents
            max_concurrent: Maximum concurrent subagents
            max_turns: Maximum turns per subagent
            timeout: Default timeout per subagent
            agent_registry: Registry of typed agent configurations
        """
        # Create registry if not provided
        self._agent_registry = agent_registry or AgentRegistry()

        self._default_config = SubagentConfig(
            model=default_model,
            system_prompt=default_system_prompt,
            tools=default_tools,
            max_turns=max_turns,
            timeout=timeout,
        )

        self._spawner = SubagentSpawner(
            default_config=self._default_config,
            max_concurrent=max_concurrent,
            agent_registry=self._agent_registry,
        )

        self._max_concurrent = max_concurrent
        self._available_tools: List[Any] = list(default_tools or [])

    @property
    def spawner(self) -> SubagentSpawner:
        """Get the subagent spawner."""
        return self._spawner

    @property
    def agent_registry(self) -> AgentRegistry:
        """Get the agent registry."""
        return self._agent_registry

    def set_available_tools(self, tools: List[Any]) -> None:
        """Set the full toolset available for subagent filtering."""
        self._available_tools = list(tools)
        self._default_config.tools = list(tools)

    def get_tools(self) -> List[ToolSpec]:
        """Return subagent tools."""
        # Build agent type description from registry
        agent_descriptions = self._agent_registry.get_descriptions()

        return [
            ToolSpec.from_function(
                self._spawn_subagent,
                name="spawn_subagent",
                description=(
                    "Launch a sub-agent to handle a specific task independently.\n\n"
                    "## When to use:\n"
                    "- Complex operations requiring multiple steps\n"
                    "- Work that benefits from isolated context\n"
                    "- Parallel-friendly tasks that don't need your current context\n"
                    "- Exploratory work (use agent='explore')\n"
                    "- Planning work (use agent='plan')\n\n"
                    "## Agent Types:\n"
                    f"{agent_descriptions}\n\n"
                    "## Important:\n"
                    "- Subagent has FRESH context (doesn't see your conversation)\n"
                    "- Provide ALL necessary context in the task description\n"
                    "- Subagent result returns to you as tool_result\n"
                    "- Use for delegation, not for simple tasks\n"
                    "- If you have 2+ independent subagent tasks, prefer spawn_parallel"
                ),
                category="subagents",
                requires_confirmation=True,
            ),
            ToolSpec.from_function(
                self._spawn_parallel,
                name="spawn_parallel",
                description=(
                    "Spawn multiple subagents to research different aspects in parallel. "
                    "Each task gets its own independent subagent. "
                    "Returns list of subagent IDs. "
                    "Prefer this over repeated spawn_subagent calls when tasks are independent."
                ),
                category="subagents",
                requires_confirmation=True,
            ),
            ToolSpec.from_function(
                self._wait_subagent,
                name="wait_subagent",
                description=(
                    "Wait for a specific subagent to complete and get its result. "
                    "Blocks until the subagent finishes."
                ),
                category="subagents",
                requires_confirmation=True,
            ),
            ToolSpec.from_function(
                self._wait_all,
                name="wait_all_subagents",
                description=(
                    "Wait for all active subagents to complete. "
                    "Returns all results."
                ),
                category="subagents",
                requires_confirmation=True,
            ),
            ToolSpec.from_function(
                self._get_result,
                name="get_subagent_result",
                description=(
                    "Get the result from a completed subagent. "
                    "Returns None if subagent is still running."
                ),
                category="subagents",
                requires_confirmation=True,
            ),
            ToolSpec.from_function(
                self._list_subagents,
                name="list_subagents",
                description="List all active and completed subagents with their status.",
                category="subagents",
            ),
            ToolSpec.from_function(
                self._cancel_subagent,
                name="cancel_subagent",
                description="Cancel a running subagent.",
                category="subagents",
            ),
            ToolSpec.from_function(
                self._run_subagent_workflow,
                name="run_subagent_workflow",
                description=(
                    "Run a structured multi-role workflow: discover -> implement -> verify"
                    " -> critic (optional), and return a merged result."
                ),
                category="subagents",
                requires_confirmation=True,
            ),
        ]

    def get_system_prompt_segment(self) -> str:
        """Return system prompt describing subagent capabilities."""
        agent_descriptions = self._agent_registry.get_descriptions()

        return f"""## Parallel Subagents

You can spawn independent subagents to handle research and complex tasks in parallel.

**Available Agent Types:**
{agent_descriptions}

**Available Tools:**
- `spawn_subagent(task, agent="default")`: Create a subagent for a task
- `spawn_parallel(tasks)`: Spawn multiple subagents at once
- `wait_subagent(id)`: Wait for a subagent to complete
- `wait_all_subagents()`: Wait for all subagents
- `get_subagent_result(id)`: Get result (non-blocking)
- `list_subagents()`: Show all subagents
- `cancel_subagent(id)`: Stop a running subagent
- `run_subagent_workflow(goal, include_critic=True)`: One-call multi-role workflow

**Configuration:**
- Max concurrent: {self._max_concurrent}
- Default timeout: {self._default_config.timeout}s
- Max turns per subagent: {self._default_config.max_turns}

**Key Principles:**
- Subagents get FRESH context (isolated from parent conversation)
- Provide ALL necessary context in the task description
- Use agent="explore" for read-only investigation
- Use agent="plan" for structured planning tasks
- Use agent="discover" for environment and codebase mapping
- Use agent="implement" for focused code changes
- Use agent="verify" for spec and test validation
- Use agent="critic" for independent quality challenge
- Use default for general task execution
- If you identify 2+ independent investigations, batch them with `spawn_parallel`
"""

    async def run_workflow(self, goal: str, include_critic: bool = True) -> Dict[str, Any]:
        """Run discover -> implement -> verify -> critic workflow."""
        discover_prompt = (
            "Goal:\n"
            f"{goal}\n\n"
            "Produce: relevant files, constraints, interfaces, and suggested verification commands."
        )
        discover = await self._spawn_and_wait(discover_prompt, agent="discover")

        implement_prompt = (
            "Goal:\n"
            f"{goal}\n\n"
            "Discovery findings:\n"
            f"{discover.get('result', discover)}\n\n"
            "Implement the required changes with focused scope."
        )
        implement = await self._spawn_and_wait(implement_prompt, agent="implement")

        verify_prompt = (
            "Goal:\n"
            f"{goal}\n\n"
            "Implementation output:\n"
            f"{implement.get('result', implement)}\n\n"
            "Verify against spec using tests/checks and return pass/fail evidence."
        )
        verify = await self._spawn_and_wait(verify_prompt, agent="verify")

        critic: Dict[str, Any] = {}
        if include_critic:
            critic_prompt = (
                "Goal:\n"
                f"{goal}\n\n"
                "Discovery:\n"
                f"{discover.get('result', discover)}\n\n"
                "Implementation:\n"
                f"{implement.get('result', implement)}\n\n"
                "Verification:\n"
                f"{verify.get('result', verify)}\n\n"
                "Provide an independent critique with specific risks and fixes."
            )
            critic = await self._spawn_and_wait(critic_prompt, agent="critic")

        return {
            "status": "completed",
            "goal": goal,
            "stages": {
                "discover": discover,
                "implement": implement,
                "verify": verify,
                "critic": critic if include_critic else {"status": "skipped"},
            },
            "summary": {
                "discover_status": discover.get("status"),
                "implement_status": implement.get("status"),
                "verify_status": verify.get("status"),
                "critic_status": critic.get("status") if include_critic else "skipped",
            },
        }

    async def _run_subagent_workflow(
        self,
        goal: str,
        include_critic: bool = True,
    ) -> Dict[str, Any]:
        """Tool wrapper for one-call workflow execution."""
        try:
            return await self.run_workflow(goal=goal, include_critic=include_critic)
        except Exception as e:
            return {"error": str(e)}

    async def _spawn_and_wait(self, task: str, agent: str) -> Dict[str, Any]:
        """Spawn subagent for a role and wait for completion."""
        subagent_id = await self._spawner.spawn(task, None, agent=agent)
        result = await self._spawner.wait(subagent_id)
        payload = result.to_dict()
        payload["agent"] = agent
        return payload

    async def _spawn_subagent(
        self,
        task: str,
        agent: str = "default",
        system_prompt: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Spawn a new subagent.

        Args:
            task: Task description for the subagent (include ALL necessary context)
            agent: Agent type from registry ("default", "explore", "plan", or custom)
            system_prompt: Optional custom system prompt (overrides agent type prompt)

        Returns:
            Dict with subagent ID and metadata
        """
        config = None
        if system_prompt:
            # Custom system prompt overrides agent type
            config = SubagentConfig(
                model=self._default_config.model,
                system_prompt=system_prompt,
                tools=self._default_config.tools,
                max_turns=self._default_config.max_turns,
                timeout=self._default_config.timeout,
            )

        try:
            subagent_id = await self._spawner.spawn(task, config, agent=agent)

            # Get agent config for metadata
            agent_config = self._agent_registry.get(agent)

            return {
                "status": "spawned",
                "id": subagent_id,
                "task": task,
                "agent": agent,
                "agent_description": agent_config.description,
            }
        except Exception as e:
            return {"error": str(e)}

    async def _spawn_parallel(
        self,
        tasks: List[str],
    ) -> Dict[str, Any]:
        """
        Spawn multiple subagents in parallel.

        Args:
            tasks: List of task descriptions

        Returns:
            Dict with list of subagent IDs
        """
        if len(tasks) > self._max_concurrent:
            return {
                "error": f"Too many tasks ({len(tasks)}). "
                f"Maximum concurrent: {self._max_concurrent}"
            }

        try:
            ids = await self._spawner.spawn_many(tasks)
            return {
                "status": "spawned",
                "count": len(ids),
                "subagents": [
                    {"id": sid, "task": task}
                    for sid, task in zip(ids, tasks)
                ],
            }
        except Exception as e:
            return {"error": str(e)}

    async def _wait_subagent(self, subagent_id: str) -> Dict[str, Any]:
        """
        Wait for a subagent to complete.

        Args:
            subagent_id: ID of the subagent

        Returns:
            Dict with subagent result
        """
        try:
            result = await self._spawner.wait(subagent_id)
            return result.to_dict()
        except ValueError as e:
            return {"error": str(e)}
        except Exception as e:
            return {"error": f"Error waiting for subagent: {e}"}

    async def _wait_all(self) -> Dict[str, Any]:
        """
        Wait for all active subagents.

        Returns:
            Dict with all results
        """
        try:
            results = await self._spawner.wait_all()
            return {
                "count": len(results),
                "results": [r.to_dict() for r in results],
            }
        except Exception as e:
            return {"error": str(e)}

    def _get_result(self, subagent_id: str) -> Dict[str, Any]:
        """
        Get result for a subagent (non-blocking).

        Args:
            subagent_id: ID of the subagent

        Returns:
            Dict with result or status
        """
        result = self._spawner.get_result(subagent_id)
        if result:
            return result.to_dict()

        status = self._spawner.get_status(subagent_id)
        if status:
            return {
                "id": subagent_id,
                "status": status.value,
                "message": "Subagent is still running" if status == SubagentStatus.RUNNING else "Unknown",
            }

        return {"error": f"Subagent '{subagent_id}' not found"}

    def _list_subagents(self) -> Dict[str, Any]:
        """
        List all subagents.

        Returns:
            Dict with subagent list
        """
        subagents = self._spawner.list_subagents()
        active = sum(1 for s in subagents if s.get("status") == "running")

        return {
            "active": active,
            "total": len(subagents),
            "subagents": subagents,
        }

    def _cancel_subagent(self, subagent_id: str) -> Dict[str, Any]:
        """
        Cancel a running subagent.

        Args:
            subagent_id: ID of the subagent

        Returns:
            Dict with cancellation status
        """
        if self._spawner.cancel(subagent_id):
            return {"status": "cancelled", "id": subagent_id}
        return {"error": f"Subagent '{subagent_id}' not found or already completed"}

    def get_state(self) -> Dict[str, Any]:
        """Get middleware state."""
        return {
            "max_concurrent": self._max_concurrent,
            "active_count": self._spawner.active_count,
            "completed_count": len(self._spawner.results),
        }

    def load_state(self, state: Dict[str, Any]) -> None:
        """Load middleware state."""
        # Note: Subagent tasks are not persisted
        self._max_concurrent = state.get("max_concurrent", self._max_concurrent)

File: vel_harness-ts/src/middleware/subagents.ts
/**
 * Subagents Middleware
 *
 * Provides tools for spawning and managing subagents.
 */

import { ToolSpec } from 'vel-ts';
import type { ModelConfig } from 'vel-ts';
import { z } from 'zod';
import { BaseMiddleware } from './base.js';
import { AgentRegistry } from '../agents/registry.js';
import { SubagentSpawner } from '../subagents/spawner.js';

// Input types
interface SpawnSubagentInput {
  task: string;
  agent?: 'default' | 'explore' | 'plan';
}

interface WaitSubagentInput {
  id: string;
  timeout?: number;
}

interface SpawnParallelInput {
  tasks: Array<{
    task: string;
    agent?: 'default' | 'explore' | 'plan';
  }>;
}

interface GetSubagentResultInput {
  id: string;
}

interface CancelSubagentInput {
  id: string;
}

interface WaitAllSubagentsInput {
  timeout?: number;
}

/**
 * Subagents middleware options
 */
export interface SubagentsMiddlewareOptions {
  /** Agent registry */
  registry: AgentRegistry;

  /** Model configuration for subagents */
  model: ModelConfig;

  /** Maximum concurrent subagents (default: 5) */
  maxConcurrent?: number;

  /** Maximum total subagents per session (default: 10) */
  maxTotalSubagents?: number;

  /** Maximum tasks per spawn_parallel call (default: 5) */
  maxParallelTasks?: number;
}

/**
 * Middleware providing subagent spawning tools
 */
export class SubagentsMiddleware extends BaseMiddleware {
  readonly name = 'subagents';

  private spawner: SubagentSpawner;
  private registry: AgentRegistry;
  private maxTotalSubagents: number;
  private maxParallelTasks: number;
  private totalSpawned = 0;

  constructor(options: SubagentsMiddlewareOptions) {
    super();
    this.registry = options.registry;
    this.maxTotalSubagents = options.maxTotalSubagents ?? 10;
    this.maxParallelTasks = options.maxParallelTasks ?? 5;
    this.spawner = new SubagentSpawner({
      registry: options.registry,
      model: options.model,
      maxConcurrent: options.maxConcurrent ?? 5,
    });
  }

  /**
   * Check if we can spawn more subagents
   */
  private canSpawn(count = 1): { allowed: boolean; reason?: string } {
    if (this.totalSpawned + count > this.maxTotalSubagents) {
      return {
        allowed: false,
        reason: `Subagent limit reached (${this.totalSpawned}/${this.maxTotalSubagents}). Complete your analysis with existing results.`,
      };
    }
    return { allowed: true };
  }

  /**
   * Reset spawn counter (call at session start)
   */
  resetSpawnCount(): void {
    this.totalSpawned = 0;
  }

  /**
   * Get the spawner (for tool injection)
   */
  getSpawner(): SubagentSpawner {
    return this.spawner;
  }

  getTools(): ToolSpec[] {
    const self = this;

    return [
      new ToolSpec<SpawnSubagentInput>({
        name: 'spawn_subagent',
        description: `Spawn a subagent for task delegation. Use sparingly - prefer to do work directly when possible.

Available agent types:
${this.registry.getDescriptions()}`,
        inputSchema: z.object({
          task: z.string().describe('Task description for the subagent'),
          agent: z
            .enum(['default', 'explore', 'plan'])
            .optional()
            .describe('Agent type to use'),
        }),
        requiresConfirmation: true,
        handler: (input: SpawnSubagentInput) => {
          // Check spawn limit
          const check = self.canSpawn(1);
          if (!check.allowed) {
            return { error: check.reason };
          }

          // spawn() is now sync - returns immediately, agent runs in background
          const result = self.spawner.spawn(input.task, input.agent ?? 'default');
          self.totalSpawned++;
          return {
            id: result.id,
            status: result.status,
            agentType: result.agentType,
            remaining: self.maxTotalSubagents - self.totalSpawned,
          };
        },
      }),

      new ToolSpec<WaitSubagentInput>({
        name: 'wait_subagent',
        description: 'Wait for a subagent to complete and get its result',
        inputSchema: z.object({
          id: z.string().describe('Subagent ID'),
          timeout: z.number().optional().describe('Timeout in milliseconds'),
        }),
        requiresConfirmation: true,
        handler: async (input: WaitSubagentInput) => {
          const result = await self.spawner.wait(input.id, input.timeout);
          return {
            id: result.id,
            status: result.status,
            result: result.result,
            error: result.error,
            duration: result.completedAt && result.startedAt
              ? result.completedAt.getTime() - result.startedAt.getTime()
              : undefined,
          };
        },
      }),

      new ToolSpec<WaitAllSubagentsInput>({
        name: 'wait_all_subagents',
        description: 'Wait for ALL running subagents to complete in PARALLEL. Use this after spawn_parallel to get all results at once.',
        inputSchema: z.object({
          timeout: z.number().optional().describe('Timeout in milliseconds (default: 300000)'),
        }),
        requiresConfirmation: true,
        handler: async (input: WaitAllSubagentsInput) => {
          const results = await self.spawner.waitAll(input.timeout);
          return {
            count: results.length,
            results: results.map((r) => ({
              id: r.id,
              task: r.task,
              agentType: r.agentType,
              status: r.status,
              result: r.result,
              error: r.error,
            })),
          };
        },
      }),

      new ToolSpec<SpawnParallelInput>({
        name: 'spawn_parallel',
        description: `Spawn multiple subagents in parallel. Max ${this.maxParallelTasks} tasks per call. Use sparingly.`,
        inputSchema: z.object({
          tasks: z.array(
            z.object({
              task: z.string().describe('Task description'),
              agent: z
                .enum(['default', 'explore', 'plan'])
                .optional()
                .describe('Agent type'),
            })
          ).max(this.maxParallelTasks),
        }),
        requiresConfirmation: true,
        handler: (input: SpawnParallelInput) => {
          // Enforce max tasks per call
          if (input.tasks.length > self.maxParallelTasks) {
            return {
              error: `Too many tasks (${input.tasks.length}). Maximum ${self.maxParallelTasks} per call.`,
            };
          }

          // Check spawn limit
          const check = self.canSpawn(input.tasks.length);
          if (!check.allowed) {
            return { error: check.reason };
          }

          // spawnMany() spawns all in parallel and returns immediately
          const results = self.spawner.spawnMany(
            input.tasks.map((t) => ({ task: t.task, agentType: t.agent }))
          );
          self.totalSpawned += results.length;

          return {
            subagents: results.map((r) => ({
              id: r.id,
              task: r.task,
              agentType: r.agentType,
              status: r.status,
            })),
            count: results.length,
            remaining: self.maxTotalSubagents - self.totalSpawned,
          };
        },
      }),

      new ToolSpec<GetSubagentResultInput>({
        name: 'get_subagent_result',
        description: 'Get the result of a completed subagent',
        inputSchema: z.object({
          id: z.string().describe('Subagent ID'),
        }),
        requiresConfirmation: true,
        handler: async (input: GetSubagentResultInput) => {
          const result = self.spawner.getResult(input.id);
          if (!result) {
            return { error: `Subagent ${input.id} not found` };
          }
          return {
            id: result.id,
            task: result.task,
            agentType: result.agentType,
            status: result.status,
            result: result.result,
            error: result.error,
          };
        },
      }),

      new ToolSpec<CancelSubagentInput>({
        name: 'cancel_subagent',
        description: 'Cancel a running subagent',
        inputSchema: z.object({
          id: z.string().describe('Subagent ID'),
        }),
        handler: async (input: CancelSubagentInput) => {
          const cancelled = self.spawner.cancel(input.id);
          return { cancelled, id: input.id };
        },
      }),
    ] as unknown as ToolSpec[];
  }

  getSystemPromptSegment(): string {
    return `# Subagents

You can delegate tasks to specialized subagents. Use them SPARINGLY - prefer direct tool use when possible.

**Limits**: Max ${this.maxTotalSubagents} subagents per session, max ${this.maxParallelTasks} per spawn_parallel call.

${this.registry.getDescriptions()}

**When to use subagents**:
- Exploring MULTIPLE INDEPENDENT codebases or directories in parallel
- When you need to search many different areas simultaneously
- Complex multi-part research that genuinely benefits from parallelism

**When NOT to use subagents**:
- Simple file reads or searches - use Read/Grep/Glob directly
- Sequential exploration - just do it yourself
- When you already have the information you need
- Iterative deepening - do one round, then decide if more is needed

**Best practices**:
- Spawn subagents ONCE for a task, then work with the results
- Don't spawn more subagents to "analyze" what previous subagents found
- Prefer fewer, well-scoped subagent tasks over many small ones
- After spawn_parallel, call wait_all_subagents to get results`;
  }
}
